src/D2L.Security.OAuth2/Keys/Default/ExpiringPublicKeyDataProvider.cs

```diff
@@ -49,17 +49,21 @@ IDateTimeProvider dateTimeProvider
 				.GetAllAsync()
 				.ConfigureAwait( false );
 
-			keys = await Task
-				.WhenAll(
-					keys.Select( key => KeyExpirationHelper( key ) ).ToArray()
-				)
-				.ConfigureAwait( false );
+			var deletes = new List<Task>();
+
+			var result = new List<JsonWebKey>();
 
-			keys = keys
-				.Where( key => key != null )
-				.ToArray();
+			foreach( var key in keys ) {
+				var keyIsExpired = await KeyIsExpiredAsync( key ).ConfigureAwait( false );

```

Comment:
1. We need this to be named -Async for the sync-to-async generator
2. We want to not call blocking methods inside `.Select`/lambdas, because its hard to keep track of, so this PR eliminates the LINQ usage.